More Python tests for materialized view and Alternator GSI feature

nyh opened the pull request:
This patch includes more tests (in Python) that I wrote while implementing the Alternator UpdateTable feature for adding a GSI to an existing table (https://github.com/scylladb/scylladb/issues/11567).

I explain each of these tests in the separate patches below, but basically they fall into two types:
1. Tests which pass with today's materialized views and Alternator GSI/LSI, and serve to ensure that whatever changes I do to the view update implementation, doesn't break corner cases that already worked.
2. Tests for the UpdateTable feature in Alternator which doesn't work today so xfail - and will need to work for #11567. We already had a few tests for this, but here I add more and improve coverage of various corner cases I discovered while implementing the featue.

I already have a working prototype for #11567 which passes all these tests. Many of these tests helped exposed various bugs in earlier versions of my code.

scylladb-promoter commented:
#### :red_circle: CI State: ABORTED

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13841/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13841/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: alternator
:small_blue_diamond: alternator/test_gsi
:small_blue_diamond: alternator/test_gsi_updatetable
:small_blue_diamond: alternator/test_lsi
:small_blue_diamond: alternator/test_metrics
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/704/)
:x: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/106/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13841/)
#### Failed Tests (1/297399):
- [test_random_failures[stop_before_becoming_raft_voter-restart_coordinator_node].debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13841/testReport/junit/topology_random_failures/test_random_failures/Tests___Unit_Tests___test_random_failures_stop_before_becoming_raft_voter_restart_coordinator_node__debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_before_becoming_raft_voter-restart_coordinator_node].debug.1)
#### Build Details:
- Duration: 8 hr 49 min
- Builder: spider1.cloudius-systems.com

piodul commented on test/alternator/test_gsi.py:
```diff
@@ -4,6 +4,10 @@
+# All the tests in this file create a GSI together with the base table - the
+# separate file test_gsi_updatetable.py has tests for the ability to use
+# UpdateTable to add or remove GSIs on existing tables.
+#
```
> This patch adds many more tests for various corner cases of
> this feature - tests developed in parallel with actually implementing
> that feature.
> 
> The source file test/alternator/test_gsi.py has already grown pretty
> large, so this patch moves all the tests related to using UpdateTable
> to add or delete a GSI - both the existing tests and the new ones -
> to a separate file: test_gsi_updatetable.py,

Please separate the code movement and addition of new tests into two commits.

nyh replied:
I can do this, but it won't be trivial, as I actually made some changes in comments and perhaps (I don't remember) other small changes in the code after the movement. 

From the **deletion** side you can see exactly what I moved (which got deleted from the old file), but you're right that you can't see exactly what I changed after the move.

piodul reviewed (commented):
Only one comment, otherwise LGTM.

I am not too familiar with Alternator, so somebody else more familiar with it should review as well - @ptrsmrn could you or somebody from your team take a look?

nyh commented:
CI failure is a flaky test - https://github.com/scylladb/scylladb/issues/21872

ptrsmrn commented on test/alternator/test_cql_rbac.py:
```diff
@@ -669,6 +672,167 @@ def test_rbac_updatetable(dynamodb, cql):
                     authorized(lambda: tab.meta.client.update_table(TableName=tab.name,
                         BillingMode='PAY_PER_REQUEST'))
 
+# Above we checked UpdateTable's operation to change BillingMode, and that
```
"Above" - that is, unless sombody doesn't insert a testcase in between.

nyh replied:
Yes, I'm writing a book here :-)
Also above here doesn't mean the immediately above test, just that we had another test for it. When I want to refer to a specific test I usually include its name exactly for the reason you asked about. I can do it here too.

nyh replied:
Fixing the text to "Above in test_rbac_updatetable() we checked". 

I know nothing guarantees that it will remain "above", but I think there's a high likelyhood that nobody will ever come to this test file and shuffle around functions randomly, so the word "above" is likely to remain correct (and if not, it can be fixed when becomes incorrected).

nyh replied:
I'm trying to do this but lacking the git skills, apparently :-( I'm still thinking how to do it without losing my mind.

nyh replied:
I managed to do it :-) Mostly using "git rebase -i" as the tool.

ptrsmrn commented on test/alternator/test_cql_rbac.py:
```diff
@@ -669,6 +672,167 @@ def test_rbac_updatetable(dynamodb, cql):
                         BillingMode='PAY_PER_REQUEST'))
 
+# Above we checked UpdateTable's operation to change BillingMode, and that
+# it requires the ALTER permissions. But UpdateTable has a more interesting
+# ability - to *create* and to *delete* a GSI from the base table. In this
```
So creating a table already with the GSI is not possible and we first need to create a table and then add a GSI to it via ALTER?

nyh replied:
No - we already supported since the first days of Alternator the `CreateTable` API to create a table with a GSI, what we're adding now is the `UpdateTable` API to modify an existing table by adding a GSI to it.
The `CreateTable` requires (unsurprisingly) the CREATE permissions, but also auto-grants all the permissions - including ALTER - to the role who was allowed to create the table, so this role is later able to add another GSI to this table if he wishes.

nyh commented:
Pushed new version with:
1. Fixed one test that I found was flaky - test_rbac_updatetable_gsi() - the test for permission denied on table **delete** needs to be a bit more clever - it needs retry but we can't retry deletion because if the first delete succeeds and deletes the GSI - we can't continue the iterations.
2. Fixed one comment as requested by a reviewer.
3. Instead of moving existing UpdateTable tests, improving them and adding more - all in one patch - now I split this change into three patches (requested by a reviewer).

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13871/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13871/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: alternator
:small_blue_diamond: alternator/test_gsi
:small_blue_diamond: alternator/test_gsi_updatetable
:small_blue_diamond: alternator/test_lsi
:small_blue_diamond: alternator/test_metrics
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/728/)
:white_check_mark: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/130/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13871/)
#### Build Details:
- Duration: 8 hr 55 min
- Builder: spider1.cloudius-systems.com

nyh commented:
Please review again and/or consider merging. I believe I implemented the reviewers' comments, and the CI passed.

nyh commented:
@scylladb/scylla-maint please consider merging this pull request. I am waiting for it to be merged to push on top of it the actual implementation of that GSI feature.

piodul commented:
Queued
